### PR TITLE
Asynchronous mode for C++ object detection sample

### PR DESCRIPTION
**Pull request by dkurt**
### This pullrequest changes

see https://github.com/opencv/opencv/pull/14516 for Python sample (will be ported into the master branch).

### CODE REVIEW

File: samples/dnn/object_detection.cpp [hunk 1]
```diff
@@ -104,44 +137,135 @@ int main(int argc, char** argv)
...
+    std::queue<Mat> processedFramesQueue;
+    QueueFPS<std::vector<Mat> > predictionsQueue;
+    std::thread processingThread([&](){
+        std::queue<std::future<Mat> > futureOutputs;
+        Mat blob;
+        while (process)
+        {
+            // Get a next frame
+            Mat frame;
+            framesQueueMutex.lock();
```
Comment (alalek) [inline samples/dnn/object_detection.cpp]: Use scope-based `std::shared_lock` or `cv::AutoLock` (with `cv::Mutex`).

File: samples/dnn/object_detection.cpp [hunk 2]
```diff
@@ -104,44 +137,135 @@ int main(int argc, char** argv)
...
+    std::thread framesThread([&](){
+        Mat frame;
+        while (process)
         {
-            waitKey();
-            break;
+            cap >> frame;
+            if (!frame.empty())
+            {
+                framesQueueMutex.lock();
```
Comment (alalek) [inline samples/dnn/object_detection.cpp]: Use scope-based `std::shared_lock` or `cv::AutoLock` (with `cv::Mutex`).

File: samples/dnn/object_detection.cpp [hunk 3]
```diff
@@ -104,44 +137,135 @@ int main(int argc, char** argv)
...
-    // Process frames.
-    Mat frame, blob;
-    while (waitKey(1) < 0)
-    {
-        cap >> frame;
-        if (frame.empty())
+    bool process = true;
+
+    // Frames capturing thread
+    std::mutex framesQueueMutex;
```
Comment (alalek) [inline samples/dnn/object_detection.cpp]: Mutex should go into `QueueFPS` class.

File: samples/dnn/object_detection.cpp [hunk 4]
```diff
@@ -41,7 +44,36 @@ void drawPred(int classId, float conf, int left, int top, int right, int bottom,
...
+    QueueFPS() : counter(0) {}
+
+    void push(const T& entry)
+    {
+        std::queue<T>::push(entry);
+        counter += 1;
+        if (counter == 1)
+        {
+            tm.reset();
+            tm.start();
```
Comment (alalek) [inline samples/dnn/object_detection.cpp]: Add comment about skipping of "warmup" first frame.

File: samples/dnn/object_detection.cpp (hunk 1 above)
Comment (dkurt) [reply samples/dnn/object_detection.cpp]: Replaced to `std::lock_guard`, thanks!

File: samples/dnn/object_detection.cpp (hunk 3 above)
Comment (dkurt) [reply samples/dnn/object_detection.cpp]: Moved all the locks into the `QueueFPS`.

**Comment by dkurt** [comment]
> Number of dropped frames?

Maybe percentage is even better.

**Comment by dkurt** [comment]
> Number of dropped frames?

That's a good idea, thank you! Ratio of dropped might be computer by division of Camera FPS over Processed FPS but the number of skipped frames is useful to know.

**Comment by alalek** [comment]
Please apply threading fixes from here: https://github.com/alalek/opencv/commit/pr14538_r

**Comment by dkurt** [comment]
Merged to https://github.com/opencv/opencv/pull/14516.

@alalek, I'll apply the patch with threads now.

**Pull request closed without merging**